### ISSUE TITLE
【BERT】BERT 系列模型的进化：RoBERTa

### ISSUE THREAD

**Issue by QingyaFan** [QingyaFan/blog#172]
整体时间线如下:

```
2018  BERT（Google）
↓ 
2019 RoBERTa(Meta)
↓
2019 XLM-ToBERTa(Meta)
↓
DeBERTa / DeBERTa V2
↓
2021 DeBERTa V3(Microsoft)
```

本文主要讲 RoBERTa 相对于 BERT。

## **RoBERTa**

RoBERTa 的论文标题是 “RoBERTa: A Robustly Optimized BERT Pretraining Approach”, 相对于 BERT,不换模型结构,只把整个训练方法优化到极致,大幅提升效果。论文中展示了效果：

> RoBERTa 在 GLUE、SQuAD、MNLI 等任务大幅超越 BERT，尤其在 句子分类 和 文本相似度 上效果更强。

其具体措施如下.

### 1. 移除 Next Sentence Prediction (NSP)任务

BERT 预训练包括两个任务:
- Masked Language Model (MLM)
- Next Sentence Prediction (判断句子 A 与 B 是否连续)

RoBERTa 中完全移除了 NSP 任务, 训练时单纯依靠 MLM. 因为 Meta 实验证明: 
- NSP 不能有效帮助模型理解句间关系.
- 去掉 NSP 可用更长的文档拼接,提高模型看到的上下文.
- 训练效率更高,更多容量用于 MLM

技术方法: 使用 Dynamic Concat 合并多个句子形成 512 上下文,不需要 A/B sentence embedding.

### 2. 使用动态 Mask (Dynamic Masking)

BERT 使用静态 Mask：数据预处理时就固定了哪些 token 被 mask，所以模型每个 epoch 看到的 mask 是一样的。

RoBERTa 使用动态 Mask：训练时在线随机生成mask，每个 epoch 每个样本的 mask 都不同。

提升的原因：
- 实际上相当于对同样的数据进行了数据增强
- 有效增加 mask 的多样性
- MLM 更难 -> 模型拟合能力更强

技术方法：在 DataLoader 里为每个 batch 动态采样 mask。

### 3. 更大规模的数据集

BERT 用的数据：
- BookCorpus（800M）
- English Wikipedia（2500M）
合计 ～ 3.3 B tokens

RoBERTa 用的数据：
- 加了 OpenWebText（35GB）
- CC-News（63M 文章）
- Stories（31B tokens）
总规模达到约 10 倍。

🎯 为什么提升：
- 语言理解能力高度依赖语料规模
- 上下文模式、长尾词更充分
- 泛化能力显著增强

🔧 技术方法：
构建更大更干净的 WebText 数据流水线并过滤低质量内容。

### 4. 更长的训练时间 + 更大 batch size

BERT：
- batch size ～ 256
- train 1M steps

RoBERTa：
- batch 可放大到 8192
- 训练更多steps
- 使用更高的 learning rate warmup

🎯 为什么提升
- 更大 batch 能更有效利用 GPU
- 使训练分布更稳定
- 更长训练时间 -> 模型更充分收敛

🔧 技术方法
Meta 使用 LAMB /Adam 大 batch 优化器策略 + LR Schedule 优化。

💡大batch 可以提升效果，但前提是做好如下这些：
1. 使用 learning rate linear scaling rule
2. 足够长的训练时间
3. warmup
4. 优化器使用 LAMB / AdamW 大 batch 配置
5. gradient clipping

如果没有做好上述工作，就会收敛差、甚至直接发散。所以，RoBERTa、GPT、LLaMA 等大模型等论文都遵循大 batch 训练范式。

### 5. 更长的序列训练

max_position_embeddings = 512

BERT 虽然最大长度 512，但训练中大量样本是 128（为了省算力）
RoBERTa 更高比例采用 训练 512 长度

🎯 为什么提升？
- 模型见到的长期依赖关系更充分
- 能更好理解长文场景（段落、文档、运维问题描述等）

🔧 技术方法：在动态 concatenation 后大量构造 512 tokens 等样本。
